File: model/data/dsp.py
"""
CW DSP envelope extraction — 4-channel orthogonal physics.

Synced from cw-ml/cw-dsp-research/dsp.py (autoresearch composite 0.8249,
v1.3.1 morse-audio AGC-calibrated SNR, eval -16 to -8 dB).

ch0: amplitude       — ±25 Hz bandpass + Hilbert + pct-norm + double sharpen
ch1: TKEO            — Teager-Kaiser energy on bandpassed signal (zero-delay)
ch2: matched filter  — 48 ms coherent IQ box — narrow BW for low-SNR
ch3: long MF         — 200 ms coherent IQ box — character-scale confidence prior

Contract:
  extract_envelope(audio, sample_rate, tone_freq) → (T, 4) float32 in [0, 1]
  T = len(audio) // 16. Dependencies: numpy, scipy only.

process_wav(wav_path, tone_freq_hz) → np.ndarray (T, 4)
  Convenience wrapper: reads WAV, runs extract_envelope.
"""
import numpy as np
import soundfile as sf
from scipy.ndimage import gaussian_filter1d, uniform_filter1d
from scipy.signal import butter, hilbert, sosfiltfilt

DSP_SAMPLE_RATE = 8000
ENVELOPE_SR = 500
DECIMATION = 16

_BP_BW_HZ = 25.0          # ch0 bandpass half-width (narrow → low-SNR)
_BP_ORDER = 1             # lowest order → shortest impulse response → sharpest edges
_TKEO_SMOOTH_MS = 30.0    # ch1: TKEO smoothing window
_MATCHED_MS = 48.0        # ch2: dit-scale IQ integration (BW~21 Hz)
_LONG_MATCHED_MS = 200.0  # ch3: character-scale IQ integration (BW~5 Hz)
_SHARPEN_GAMMA = 8.0      # applied twice (effective γ≈64 via composition)


def extract_envelope(audio: np.ndarray, sample_rate: int = DSP_SAMPLE_RATE,
                     tone_freq: float = 600.0) -> np.ndarray:
    n = len(audio)
    n_out = n // DECIMATION
    audio64 = audio.astype(np.float64)

    lo = max(tone_freq - _BP_BW_HZ, 1.0)
    hi = min(tone_freq + _BP_BW_HZ, sample_rate / 2 - 1)
    sos_bp = butter(_BP_ORDER, [lo, hi], btype="bandpass", fs=sample_rate, output="sos")
    bp = sosfiltfilt(sos_bp, audio64)

    ch0 = _ch0_amplitude(bp, n_out)
    ch1 = _tkeo(bp, sample_rate, n_out)
    ch2 = _matched(audio64, sample_rate, tone_freq, n_out, _MATCHED_MS)
    ch3 = _matched(audio64, sample_rate, tone_freq, n_out, _LONG_MATCHED_MS)

    return np.stack([ch0, ch1, ch2, ch3], axis=1).astype(np.float32)


def process_wav(wav_path: str, tone_freq_hz: float,
                sample_rate: int = DSP_SAMPLE_RATE) -> np.ndarray:
    """Read a WAV file and return the 4-channel DSP envelope (T, 4)."""
    audio, sr = sf.read(wav_path, dtype="float32")
    if sr != sample_rate:
        raise ValueError(f"Expected {sample_rate} Hz, got {sr} in {wav_path}")
    if audio.ndim > 1:
        audio = audio[:, 0]
    return extract_envelope(audio, sample_rate, tone_freq_hz)


def _ch0_amplitude(bp: np.ndarray, n_out: int) -> np.ndarray:
    mag = np.abs(hilbert(bp))
    mag = gaussian_filter1d(mag, sigma=4.0, mode="reflect")
    env = _decimate(mag, DECIMATION)[:n_out]
    env = _normalize(env)
    env = np.clip((env - 0.05) / 0.76, 0.0, 1.0)
    env = _sharpen(env, _SHARPEN_GAMMA)
    return _sharpen(env, _SHARPEN_GAMMA)


def _tkeo(bp: np.ndarray, sample_rate: int, n_out: int) -> np.ndarray:
    psi = np.zeros_like(bp)
    psi[1:-1] = bp[1:-1] ** 2 - bp[:-2] * bp[2:]
    psi = np.maximum(psi, 0.0)
    win = max(3, int(_TKEO_SMOOTH_MS / 1000.0 * sample_rate))
    psi = uniform_filter1d(psi, size=win, mode="reflect")
    env = _decimate(psi, DECIMATION)[:n_out]
    return _normalize(env)


def _matched(audio: np.ndarray, sample_rate: int, tone_freq: float,
             n_out: int, duration_ms: float) -> np.ndarray:
    t = np.arange(len(audio)) / sample_rate
    I = audio * np.cos(2.0 * np.pi * tone_freq * t)
    Q = audio * (-np.sin(2.0 * np.pi * tone_freq * t))
    win = max(3, int(duration_ms / 1000.0 * sample_rate))
    I_mf = uniform_filter1d(I, size=win, mode="reflect")
    Q_mf = uniform_filter1d(Q, size=win, mode="reflect")
    mag = np.sqrt(I_mf ** 2 + Q_mf ** 2)
    env = _decimate(mag, DECIMATION)[:n_out]
    return _normalize(env)


def _decimate(x: np.ndarray, factor: int) -> np.ndarray:
    n = len(x) // factor * factor
    return x[:n].reshape(-1, factor).mean(axis=1)


def _normalize(env: np.ndarray, lo_pct: float = 17.0, hi_pct: float = 88.0) -> np.ndarray:
    lo = float(np.percentile(env, lo_pct))
    hi = float(np.percentile(env, hi_pct))
    denom = max(hi - lo, 1e-10)
    return np.clip((env - lo) / denom, 0.0, 1.0)


def _sharpen(x: np.ndarray, gamma: float) -> np.ndarray:
    xg = x ** gamma
    return xg / (xg + (1.0 - x) ** gamma + 1e-12)

File: cw-decode/apps/src/inference/dsp.ts
// Port of cw-ml/cw-dsp-research/dsp.py — 4-channel envelope extraction.
//
// ch0: amplitude       — ±25 Hz bandpass + Hilbert + pct-norm + gentle sharpen
// ch1: TKEO            — Teager-Kaiser energy on bandpassed signal
// ch2: matched filter  — 48 ms coherent IQ box (dit-scale, BW ~21 Hz)
// ch3: long matched    — 200 ms coherent IQ box (character-scale, BW ~5 Hz)
//
// Input audio must be at DSP_SAMPLE_RATE. Output is (T, 4) at ENVELOPE_SR, with T = floor(len/16).

import { fft, ifft, nextPow2 } from './fft'

export const DSP_SAMPLE_RATE = 8000
export const ENVELOPE_SR = 500
export const DECIMATION = 16

const BP_BW_HZ = 25.0
const TKEO_SMOOTH_MS = 30.0
const MATCHED_MS = 48.0
const LONG_MATCHED_MS = 200.0
const SHARPEN_GAMMA = 8.0

// ---------- Biquad filter (RBJ BPF), forward-backward for ~zero phase ----------

function rbjBandpass(f0: number, bw: number, fs: number): [number, number, number, number, number] {
  const w0 = (2 * Math.PI * f0) / fs
  const cosW = Math.cos(w0)
  const sinW = Math.sin(w0)
  // BW in Hz → Q
  const Q = f0 / bw
  const alpha = sinW / (2 * Q)
  const b0 = alpha
  const b1 = 0
  const b2 = -alpha
  const a0 = 1 + alpha
  const a1 = -2 * cosW
  const a2 = 1 - alpha
  return [b0 / a0, b1 / a0, b2 / a0, a1 / a0, a2 / a0]
}

function biquad(x: Float64Array, coefs: [number, number, number, number, number]): Float64Array {
  const [b0, b1, b2, a1, a2] = coefs
  const n = x.length
  const out = new Float64Array(n)
  let x1 = 0, x2 = 0, y1 = 0, y2 = 0
  for (let i = 0; i < n; i++) {
    const y = b0 * x[i] + b1 * x1 + b2 * x2 - a1 * y1 - a2 * y2
    x2 = x1; x1 = x[i]
    y2 = y1; y1 = y
    out[i] = y
  }
  return out
}

function filtfilt(x: Float64Array, coefs: [number, number, number, number, number]): Float64Array {
  const fwd = biquad(x, coefs)
  const rev = new Float64Array(fwd.length)
  for (let i = 0; i < fwd.length; i++) rev[i] = fwd[fwd.length - 1 - i]
  const back = biquad(rev, coefs)
  const out = new Float64Array(back.length)
  for (let i = 0; i < back.length; i++) out[i] = back[back.length - 1 - i]
  return out
}

export function bandpass(audio: Float64Array, fs: number, f0: number, halfBw: number): Float64Array {
  const bw = 2 * halfBw
  const coefs = rbjBandpass(f0, bw, fs)
  return filtfilt(audio, coefs)
}

// ---------- Hilbert transform magnitude via FFT ----------

export function hilbertMag(x: Float64Array): Float64Array {
  const n = x.length
  const N = nextPow2(n)
  const re = new Float64Array(N)
  const im = new Float64Array(N)
  for (let i = 0; i < n; i++) re[i] = x[i]

  fft(re, im)

  // Multiplier: 1 for DC and Nyquist, 2 for 1..N/2-1, 0 for N/2+1..N-1
  for (let k = 1; k < N / 2; k++) {
    re[k] *= 2
    im[k] *= 2
  }
  for (let k = N / 2 + 1; k < N; k++) {
    re[k] = 0
    im[k] = 0
  }

  ifft(re, im)

  const mag = new Float64Array(n)
  for (let i = 0; i < n; i++) {
    mag[i] = Math.hypot(re[i], im[i])
  }
  return mag
}

// ---------- Smoothing ----------

export function gaussianFilter1d(x: Float64Array, sigma: number): Float64Array {
  const radius = Math.max(1, Math.ceil(sigma * 4))
  const size = 2 * radius + 1
  const kernel = new Float64Array(size)
  let sum = 0
  for (let i = 0; i < size; i++) {
    const d = i - radius
    kernel[i] = Math.exp((-d * d) / (2 * sigma * sigma))
    sum += kernel[i]
  }
  for (let i = 0; i < size; i++) kernel[i] /= sum
  return convolveReflect(x, kernel)
}

export function uniformFilter1d(x: Float64Array, size: number): Float64Array {
  // scipy.ndimage.uniform_filter1d: odd size, centered; mode=reflect
  if (size < 1) size = 1
  const half = Math.floor(size / 2)
  const n = x.length
  const out = new Float64Array(n)
  const inv = 1 / size
  for (let i = 0; i < n; i++) {
    let sum = 0
    for (let k = -half; k < size - half; k++) {
      const idx = reflectIdx(i + k, n)
      sum += x[idx]
    }
    out[i] = sum * inv
  }
  return out
}

function reflectIdx(i: number, n: number): number {
  if (n === 1) return 0
  const period = 2 * n - 2
  let k = i % period
  if (k < 0) k += period
  return k < n ? k : period - k
}

function convolveReflect(x: Float64Array, kernel: Float64Array): Float64Array {
  const n = x.length
  const kn = kernel.length
  const half = Math.floor(kn / 2)
  const out = new Float64Array(n)
  for (let i = 0; i < n; i++) {
    let acc = 0
    for (let k = 0; k < kn; k++) {
      const idx = reflectIdx(i + k - half, n)
      acc += x[idx] * kernel[k]
    }
    out[i] = acc
  }
  return out
}

// ---------- Decimate (mean pool by factor) ----------

export function decimate(x: Float64Array, factor: number): Float64Array {
  const nOut = Math.floor(x.length / factor)
  const out = new Float64Array(nOut)
  for (let i = 0; i < nOut; i++) {
    let sum = 0
    const base = i * factor
    for (let k = 0; k < factor; k++) sum += x[base + k]
    out[i] = sum / factor
  }
  return out
}

// ---------- Normalize / sharpen ----------

export function percentileNormalize(x: Float64Array, loPct = 17.0, hiPct = 88.0): Float64Array {
  const sorted = Float64Array.from(x).sort()
  const lo = percentile(sorted, loPct)
  const hi = percentile(sorted, hiPct)
  const denom = Math.max(hi - lo, 1e-10)
  const out = new Float64Array(x.length)
  for (let i = 0; i < x.length; i++) {
    const v = (x[i] - lo) / denom
    out[i] = v < 0 ? 0 : v > 1 ? 1 : v
  }
  return out
}

function percentile(sorted: Float64Array, pct: number): number {
  // Match numpy 'linear' interpolation
  if (sorted.length === 0) return 0
  const n = sorted.length
  const rank = (pct / 100) * (n - 1)
  const lo = Math.floor(rank)
  const hi = Math.ceil(rank)
  if (lo === hi) return sorted[lo]
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (rank - lo)
}

export function sharpen(x: Float64Array, gamma: number): Float64Array {
  const out = new Float64Array(x.length)
  for (let i = 0; i < x.length; i++) {
    const v = x[i]
    const xg = Math.pow(v, gamma)
    out[i] = xg / (xg + Math.pow(1 - v, gamma) + 1e-12)
  }
  return out
}

function clipOffsetScale(x: Float64Array, offset: number, scale: number): Float64Array {
  const out = new Float64Array(x.length)
  for (let i = 0; i < x.length; i++) {
    const v = (x[i] - offset) / scale
    out[i] = v < 0 ? 0 : v > 1 ? 1 : v
  }
  return out
}

// ---------- Channels ----------

function ch0Amplitude(bp: Float64Array, nOut: number): Float64Array {
  const mag = hilbertMag(bp)
  const smooth = gaussianFilter1d(mag, 4.0)
  const dec = decimate(smooth, DECIMATION).subarray(0, nOut) as Float64Array
  let env = percentileNormalize(dec)
  env = clipOffsetScale(env, 0.05, 0.76)
  env = sharpen(env, SHARPEN_GAMMA)
  env = sharpen(env, SHARPEN_GAMMA)
  return env
}

function tkeo(bp: Float64Array, fs: number, nOut: number): Float64Array {
  const psi = new Float64Array(bp.length)
  for (let i = 1; i < bp.length - 1; i++) {
    const v = bp[i] * bp[i] - bp[i - 1] * bp[i + 1]
    psi[i] = v > 0 ? v : 0
  }
  const win = Math.max(3, Math.round(TKEO_SMOOTH_MS / 1000 * fs))
  const smooth = uniformFilter1d(psi, win)
  const dec = decimate(smooth, DECIMATION).subarray(0, nOut) as Float64Array
  return percentileNormalize(dec)
}

function matched(audio: Float64Array, fs: number, toneFreq: number, nOut: number, durationMs: number): Float64Array {
  const n = audio.length
  const I = new Float64Array(n)
  const Q = new Float64Array(n)
  const twoPi = 2 * Math.PI * toneFreq / fs
  for (let i = 0; i < n; i++) {
    const phase = twoPi * i
    I[i] = audio[i] * Math.cos(phase)
    Q[i] = audio[i] * -Math.sin(phase)
  }
  const win = Math.max(3, Math.round(durationMs / 1000 * fs))
  const Imf = uniformFilter1d(I, win)
  const Qmf = uniformFilter1d(Q, win)
  const mag = new Float64Array(n)
  for (let i = 0; i < n; i++) mag[i] = Math.hypot(Imf[i], Qmf[i])
  const dec = decimate(mag, DECIMATION).subarray(0, nOut) as Float64Array
  return percentileNormalize(dec)
}

// ---------- Public API ----------

export function extractEnvelope(
  audio: Float32Array | Float64Array,
  sampleRate: number = DSP_SAMPLE_RATE,
  toneFreq: number = 700,
): Float32Array {
  if (sampleRate !== DSP_SAMPLE_RATE) {
    throw new Error(`expected ${DSP_SAMPLE_RATE} Hz audio, got ${sampleRate}`)
  }
  const audio64 = audio instanceof Float64Array ? audio : Float64Array.from(audio)
  const n = audio64.length
  const nOut = Math.floor(n / DECIMATION)

  const loHz = Math.max(toneFreq - BP_BW_HZ, 1)
  const hiHz = Math.min(toneFreq + BP_BW_HZ, sampleRate / 2 - 1)
  const center = (loHz + hiHz) / 2
  const halfBw = (hiHz - loHz) / 2
  const bp = bandpass(audio64, sampleRate, center, halfBw)

  const ch0 = ch0Amplitude(bp, nOut)
  const ch1 = tkeo(bp, sampleRate, nOut)
  const ch2 = matched(audio64, sampleRate, toneFreq, nOut, MATCHED_MS)
  const ch3 = matched(audio64, sampleRate, toneFreq, nOut, LONG_MATCHED_MS)

  // Interleave as (T, 4)
  const out = new Float32Array(nOut * 4)
  for (let i = 0; i < nOut; i++) {
    out[i * 4 + 0] = ch0[i]
    out[i * 4 + 1] = ch1[i]
    out[i * 4 + 2] = ch2[i]
    out[i * 4 + 3] = ch3[i]
  }
  return out
}
